posts/virtual_environments.md

```diff
@@ -0,0 +1,126 @@
+
+Thus, for example, adding a new column is not considered to be a breaking change but removing or modifying the existing one - is.
+
+![Figure 4: Breaking vs. non-breaking changes](virtual_environments/change_categorization.png)
+*Figure 4: Breaking vs. non-breaking changes.*
+
+This way SQLMesh can automatically reduce the amount of necessary recomputation to a minimum while guaranteeing correctness. No user intervention needed.
+
+Presently this approach doesn’t extend beyond looking at each model as a whole. Soon, however, we’ll be able to categorize changes per individual column instead, thanks to the platform’s support for column-level lineage. The platform will achieve an even finer balance between correctness and efficiency, since changes like removing a column that is not referenced downstream will no longer be categorized as a “breaking”.
+
+![Figure 5: Column-level change categorization](virtual_environments/partial_breaking.png)
+*Figure 5: Column-level change categorization.*
+
+#### Physical Layer
+When a model is modified directly or indirectly, a new snapshot which captures the change gets associated with its own unique table (or a view) in the **physical layer**, the name of which is based on the snapshot's fingerprint. Thus different versions of the same model can coexist and be evaluated at the same time without overriding each others' outputs. The platform keeps track of missing data intervals for each individual snapshot and automatically [backfills](https://sqlmesh.readthedocs.io/en/stable/concepts/plans/#backfilling) existing data gaps during the plan application.
+
+Model snapshots alone are not enough, however, since various dataset versions still need to be exposed in different environments as well as be able to swap these versions on the fly without requiring changes to existing workflows. Both of these items are handled in the **virtual layer**.
+
+### Virtual Layer
+When a downstream consumer or a workflow wants to access data produced by a model they do this by querying a view that is part of the **virtual layer**. Datasets generated by model snapshots are never accessed directly which allows SQLMesh update these views to point to different dataset versions without affecting downstream. Sort of like using a pointer or a reference in programming.
+
+If a snapshot is not referenced by any views in the **virtual layer**, it gets automatically garbage collected by the janitor process after a configurable period of time.
```

Comment:
Should we add that snapshots are kept around to make it easy to revert back to previous behavior without incurring cost?